Add versioning options (#69)
vsoch/docsy-jekyll: _docs/versioning.md

---
title: Versioning your Docs
description: How to turn on and use versioning
---

# Versioning

> Versioning documentation is a useful feature that can be achieved using Docsy Jekyll. It allows you to take snapshots of your documentation that may align with a specific version of a product or service, or you may wish to just keep a historical view of the changes that have been made to your documents over time.

Reviewer: Let's create a summary after this initial paragraph to quickly explain what adding a version is. E.g., 

To add a new version, you will need to:

1. Copy the _docs content into a subfolder of _docs/Archive (e.g., named by the version)
2. Add a corresponding table of contents to `_data/versions` and to the lookup
3. Test your site locally to ensure it works as expected.

Author: Added to the documentation as requested and is part of the latest commit.

{% include alert.html type="info" title="Note: This only allows you to version documents inside the _docs folder, it does not allow you to version any files outside of that, such as Pages, Posts etc." %}

## Implementation

> The way in which versioning is achieved is by duplicating your whole document directory and placing it inside an 'archive' subdirectory with the name of that version. The following illustrates what the base document root looks like without another version for this Docsy Jekyll site.

Reviewer: ```suggestion
The way in which versioning is achieved is by duplicating your whole document directory and placing it inside an 'archive' subdirectory with the name of that version. The following illustrates what the base document root looks like *without* another version for this Docsy Jekyll site.
```

```
document root
│      
└─── _data
└─── _docs
        └─── extras
        └─── subfolder
        example-page.md
        getting-started.md
        subfolder.md
        versioning.md
```

> To create a version of this documentation we first create a folder with the version name we want to use inside of the 'archive' folder and copy all the files and folders from our root _docs folder into that new folder, we therefore end up with a structure such as below where we moved the older documentation into the folder 'Previous'. Once we have done that we effectively have taken a snapshot of our documents at a point in time and will no longer maintain these. If you look at the structure of this site you will see the 'Previous' folder, which now holds our previous documentation, does not contain the versioning.md markdown file as this is new in the current release.

Reviewer: ```suggestion
To create a version of this documentation we first create a folder with the version name we want to use inside of the 'archive' folder and copy all the files and folders from our root _docs folder into that new folder. That might look like this:

<insert code block here - can't do alongside suggestion!>

After we do this, we end up with a structure shown below - the older documentation is moved into the folder 'Previous'. This step effectively takes a snapshot of our documents at a point in time and the expectation is that we will no longer maintain these pages. If you look at the structure of this site you will see the 'Previous' folder, which now holds our previous documentation, does not contain the versioning.md markdown file as this is new in the current release. This means that if you look at this page on the current site and then choose "Previous" from the version tab, you will get a 404 (not found).
```
Here is the code block

```bash
# Our last release is version "Previous" (could also be 1.0.0 or semver), we are pinning the docs to this version`
mkdir -p _docs/Archive/Previous`
# copy all content under _docs except Archive into this new folder.`
cp -r _docs/!(Archive) _docs/Archive/Previous`
```

<pre><code>
document root
│      
└─── _data
└─── _docs
        └─── extras
        └─── subfolder
        └─── Archive
              └─── <span style="color:blue;"><b>Previous</b></span>
                  └─── extras
                  └─── subfolder
                  example-page.md
                  getting-started.md
                  subfolder.md
        example-page.md
        getting-started.md
        subfolder.md
        versioning.md
</code></pre>

You would now continue to add, update or remove markdown files from the main _docs directory in order to work on the current version of your documentation. 

Once the versioning has been correctly configured and you have created a snapshot, you can access those documents by adding the version to the url to access that archived version (e.g. /docs/Archive/Previous/getting-started). 

To enable the users to interact with these versioned documents through the UI there needs to be some updates to the _config.yml file to setup and enable it. The following is a snippet from the _config.yml file specific to the versioning options that are available.

```yml
version_params:
  version_menu: "Release"

>   version_dir: Archive

Reviewer: Perhaps put a note also that this needs to correspond to the folder name too?

Author: I did add some additional information about that parameter in the table below that config when I first added it. Is that sufficient or do you think it needs to be more explicit than that?

Reviewer: I think it's sufficient!

{% include alert.html type="warning" title="Note: The versions listed must have the same name as the folder name in which that version of the documentation resides otherwise jekyll it will not find it! The exception is the version 'current' which does not have a separate folder and is used to inform jekyll to utilise the base _docs folder." %}

### Version Keywords
There are six special keywords that can be used in the versions config.
* main
* alpha
* beta
* rc
* pre
* *current*

The value **main**, is used to display the latest development version of the documentation, a banner is used to denote that the code and documentation maybe unstable and that they are not guaranteed to be correct or up to date and could change at any time.

The value of **alpha**, **beta**, **rc** and **pre** are used to denote that the documentation is a pre-production version and as such a banner is displayed announcing that the code/documentation maybe unstable and the content in the pages are still work in progress so may not be completely up to date or correct.

> If you wish to utilise this feature then you must create folders within your _docs/*Archive* directory with the names of those you wish to use and place your documentation inside it/them.

Reviewer: ```suggestion
If you wish to utilize this feature then you must create folders within your _docs/*Archive* directory with the names of those you wish to use and place your documentation inside it/them.
```
Sorry I'm a terrible person that uses U.S English! :laughing:

> The value of **current** is used to align with the latest version of your documentation in your base _docs folder so there **doesn't** need to be a specific folder created and named 'current' to serve them. The name 'current', for the value of 'latest' in the previous example, can be changed to whatever you wish to name the current release but be aware that the version set in the _config.yml file also needs to match it! See the example below where we have called the latest release v3.2 rather than 'current' and we also have a version exactly named v3.2 too, as mentioned this version will point to your base _docs directory. 

Reviewer: You write really good documentation! This might be the best doc contribution I've had in... 13 years!

Author: Thanks for the compliment, I'm glad that my contributions are worthwhile! :-)

```yml
title: Documentation
  url: docs
  links:
    - title: "Getting Started"
      url: "docs/getting-started"
      children:
        - title: Features
          url: "docs/getting-started#getting-started"
        - title: Development
          url: "docs/getting-started#development"
        - title: Customization
          url: "docs/getting-started#customization"
    - title: Versioning
      url: "docs/versioning"
    - title: "About"
      url: "about"
    - title: "News"
      url: "news"
- title: "Extras"
  url: "docs/extras"
  links:
    - title: Quizzes
      url: "docs/extras/example-quiz"
    - title: Tags Page
      url: "tags"
```

Now not every version of your documentation will have a change in structure and therefore require an update to toc.yml. To remove this overhead a mapping file is used to tell Docsy Jekyll which TOC it should use for a particular version,
the benefit of this is that if the structure of your site does not change then you do not need to create an additional toc.yml file.

> You will have a toc-mapping.yml file in your _data directory and the contents will contain the code below to map a particular version with a particular TOC file to use. You need to update this file for each version you create even if you do not need to create a new TOC file.

Reviewer: For my note above - I think this mapping would be fine to leave at the top level, and (per the same reason of many versions) we should have a subfolder for the archived ones.

Author: I did leave the mapping file at the top level and made the versions available in a user configurable subfolder.

```yml
# This file can be used to explicitly map a release to a specific table-of-contents
# (TOC). You'll want to use this after any revamps to information architecture, to ensure

> # that the navigation for older versions still work.

Reviewer: ```suggestion
# that the navigation for older versions still works.
```

{% include alert.html type="info" title="Note: This only allows you to version documents inside the _docs folder, it does not allow you to version any files outside of that, such as Pages, Posts etc." %}

## Implementation

The way in which versioning is achieved is by duplicating your whole document directory and placing it inside an 'archive' subdirectory with the name of that version. The following illustrates what the base document root looks like *without* another version for this Docsy Jekyll site.

```
document root
│      
└─── _data
└─── _docs
        └─── extras
        └─── subfolder
        example-page.md
        getting-started.md
        subfolder.md
        versioning.md
```
To create a version of this documentation we first create a folder with the version name we want to use inside of the 'archive' folder and copy all the files and folders from our root _docs folder into that new folder. That might look like this:

> <insert code block here - can't do alongside suggestion!>

Reviewer: oh oops! So I had a code block in my suggested comment (and the only reason I couldn't include it is because it wouldn't fit within the suggestion) so I think we need to add it here. Sorry I wasn't clear!

The goal was to show the reader a quick way to do this copy.

Reviewer: Here it is again:

```bash
# Our last release is version "Previous" so we are pinning the docs to this version`
mkdir -p _docs/Archive/Previous
# copy all content under _docs except Archive into this new folder.
cp -r _docs/!(Archive) _docs/Archive/Previous`
```

Author: Hmmm, odd as I added that to the instructions! Let me check and recommit if needs be...

Author: Somehow some files were missed, sorry about that. You should now have all the changes you mentioned.

## Table of Contents Handling (TOC)

As with the documents, your TOC contents also need to be updated so they correctly point to the document versions you are viewing.

You will need to update the toc file as the structure of your documentation/toc changes over time but as we have taken a snapshot of our documentation so we need to do the same thing for our toc so it reflects the document structure at that point in time too. You will see that for this site we have the versions of Current and Previous and you will see that the Current version has all the information about versioning which didn't exist in the prior version.

To enable the TOC to change based on the version, we have to create a new toc file and name it {version}-toc.yml and it must be placed in a subdirectory of _data whose name is specified by the tocversion_dir parameter in the _config.yml file (for this site it can be found in the _data/versions subdirectory). We have created previous-toc.yml in the _data/versions directory that has the TOC structure for the prior version where we did not have the versioning.md file present. 

> To create this version of the toc file file we can use the following commands

Reviewer: Good thinking!
